File: 09_Generics/02_variance/java/ListVariance.java
import java.util.*;

class A {}
class B extends A {}
class C extends B {}

public class ListVariance {

	private static int processAs(ArrayList<? extends A> aas) {
		int count = 0;
		for (int i = 0; i < aas.size(); ++i) {
			count += 1;
			A a = aas.get(i);
		}
		return count;
	}

	private static int processCs(ArrayList<? super C> acs) {
		int count = 0;
		for (int i = 0; i < acs.size(); ++i) {
			count += 1;
			// C c = acs.get(i); // allowed?
		}
		return count;
	}

	public static void main(String[] args) {
			
		ArrayList<B> bs = new ArrayList<B>() { {
			add(new B()); add(new B()); add(new B());
		} };

		int na = processAs(bs);
		int nc = processCs(bs);
		
		System.out.println(na);
		System.out.println(nc);
	}
}

File: 09_Generics/02_variance/kotlin/ListVariance.kt
open class A
open class B : A()
open class C : B()

fun <T> use(unused : T) : T = unused

fun processAs(aas: MutableList<out A>): Int {
    var count = 0
    for (i in aas.indices) {
        count += 1
        val a: A = aas.get(i)
        use(a)
        //aas[i] = A()   // allowed?
    }
    return count
}

fun processCs(acs: MutableList<in C>): Int {
    var count = 0
    for (i in acs.indices) {
        count += 1
        acs[i] = C()
        //val c : C = acs.get(i); // allowed?
    }
    return count
}

fun main() {
    val bs = mutableListOf( B(), B(), B() )

    val na = processAs(bs)
    val nc = processCs(bs)

    println(na)
    println(nc)
}
